src/lib.rs

```diff
@@ -4,26 +4,329 @@
+    // otherwise there is no pending payout
+    next_payment_at_block: Option<BlockNumber>,
+
+    // When set, will be the basis for automatically setting next payment,
+    // otherwise any upcoming payout will be a one off.
+    payout_interval: Option<BlockNumber>,
+
+    // stats
+    // Total payout received in this relationship
+    total_reward_received: Balance,
+
+    // Total payout failed in this relationship
+    total_reward_missed: Balance,
+}
 
 decl_storage! {
     trait Store for Module<T: Trait> as RecurringReward {
+        Recipients get(recipients): map RecipientId => Recipient<BalanceOf<T>>;
+
+        NextRecipientId get(next_recipient_id): RecipientId = FIRST_RECIPIENT_ID;
 
+        RewardRelationships get(reward_relationships): map RewardRelationshipId => RewardRelationship<T::AccountId, BalanceOf<T>, T::BlockNumber>;
+
+        NextRewardRelationshipId get(next_reward_relationship_id): RewardRelationshipId = FIRST_REWARD_RELATIONSHIP_ID;
     }
 }
 
 decl_module! {
     pub struct Module<T: Trait> for enum Call where origin: T::Origin {
 
-        fn on_finalize(now: T::BlockNumber) {}
+        fn on_finalize(now: T::BlockNumber) {
+            Self::do_payouts(now);
+        }
     }
 }
 
-impl<T: Trait> Module<T> {}
+#[derive(Eq, PartialEq, Debug)]
+pub enum RewardsError {
+    RecipientNotFound,
+    RewardSourceNotFound,
+    BlockNumberInPast,
+    RewardRelationshipNotFound,
+}
+
+pub enum NextPaymentSchedule<BlockNumber> {
+    Absolute(BlockNumber),
+    Relative(BlockNumber),
+}
+
+impl<T: Trait> Module<T> {
+    /* Adds a new Recipient recipient to recipients, with identifier equal to nextRecipientId, which is also incremented, and returns the new recipient identifier. */
+    pub fn add_recipient() -> RecipientId {
+        let next_id = Self::next_recipient_id();
+        NextRecipientId::mutate(|n| {
+            *n += 1;
+        });
+        <Recipients<T>>::insert(&next_id, Recipient::default());
+        next_id
+    }
+
+    /// Removes a mapping from reward_recipients based on the given identifier.
+    pub fn remove_recipient(id: RecipientId) {
+        <Recipients<T>>::remove(&id);
+    }
+
+    // Adds a new RewardRelationship to rewardRelationships, for a given source, recipient, account, etc., with identifier equal to current nextRewardRelationshipId. Also increments nextRewardRelationshipId.
+    pub fn add_reward_relationship(
+        mint_id: MintId,
+        recipient: RecipientId,
+        account: T::AccountId,
+        amount_per_payout: BalanceOf<T>,
+        next_payment_schedule: Option<NextPaymentSchedule<T::BlockNumber>>,
+        payout_interval: Option<T::BlockNumber>,
+    ) -> Result<RewardRelationshipId, RewardsError> {
+        ensure!(
+            <minting::Module<T>>::mint_exists(mint_id),
+            RewardsError::RewardSourceNotFound
+        );
+
+        let next_payment_at_block = match next_payment_schedule {
```

Comment:
This whole section is too complex, still, am not 100% sure it actually is correct. Let's favor the simplicity of having the user just provide whatever  `next_payment_in_block` they want. `NextPaymentSchedule` is just making this routine do much more than it should in an effort to be helpful.

Reply:
agreed, although I'm sure its correct, it perhaps belongs in the client module.